kernel/aster-nix/src/syscall/dup.rs

```diff
@@ -2,33 +2,80 @@
 
 use super::SyscallReturn;
 use crate::{
-    fs::file_table::{FdFlags, FileDesc},
+    fs::file_table::{FdFlags, FileDesc, FileTable},
     prelude::*,
+    process::ResourceType,
 };
 
 pub fn sys_dup(old_fd: FileDesc) -> Result<SyscallReturn> {
     debug!("old_fd = {}", old_fd);
 
     let current = current!();
-    let mut file_table = current.file_table().lock();
-    let file = file_table.get_file(old_fd)?.clone();
-    // The two file descriptors do not share the close-on-exec flag.
-    let new_fd = file_table.insert(file, FdFlags::empty());
+    let (old_file, new_fd) = {
+        let mut file_table: MutexGuard<FileTable> = current.file_table().lock();
+        let file = file_table.get_file(old_fd)?.clone();
+        // Generate an unsed file descriptor.
+        let new_fd = file_table.get_min_free_fd(0) as FileDesc;
```

Comment:
Both the `dup`-family system calls and the `fcntl` system call need to implement the dup semantics.

```rust
pub fn sys_fcntl(fd: FileDesc, cmd: i32, arg: u64) -> Result<SyscallReturn> {
    let fcntl_cmd = FcntlCmd::try_from(cmd)?;
    debug!("fd = {}, cmd = {:?}, arg = {}", fd, fcntl_cmd, arg);
    match fcntl_cmd {
        FcntlCmd::F_DUPFD => {
            let current = current!();
            let mut file_table = current.file_table().lock();
            let new_fd = file_table.dup(fd, arg as FileDesc, FdFlags::empty())?;
            Ok(SyscallReturn::Return(new_fd as _))
        }
        FcntlCmd::F_DUPFD_CLOEXEC => {
            let current = current!();
            let mut file_table = current.file_table().lock();
            let new_fd = file_table.dup(fd, arg as FileDesc, FdFlags::CLOEXEC)?;
            Ok(SyscallReturn::Return(new_fd as _))
        }
```

We want to maximize code reuse and avoid code duplication. So either the `fcntl` system call reuses the code of `dup` system call or both the system calls should reuse a `dup` method provided by `FileTable`. I think both approaches are acceptable. If the former approach is chosen, then `FileTable::dup` should be removed.